### ISSUE TITLE
App crashes due to EXC_BAD_ACCESS(code=EXCI386_GPFLT), when a track is playing and user presses back button

### ISSUE THREAD

**Issue by charleshkang** [spotify/ios-sdk#615]
Triggered by Thread: 14 (com.spotify.AudioStreaming)
Exception Type: EXC_BAD_ACCESS(code=EXCI386_GPFLT)

This is a very similar issue to https://github.com/spotify/ios-sdk/issues/293

This happens when I press on a playlist, then press back, then press on a playlist again, then press back. The app will crash and I'm not sure why. 

It can also happen on the 1st back as well. Not sure what this behavior means. 

Here's a gif of what's happening, sorry for it being slow! 
![spotifycrash1](https://cloud.githubusercontent.com/assets/12476080/13484085/983c5978-e0c8-11e5-9384-345dbdf446b1.gif)

**Comment by jakereps** [comment]
Try wrapping your `playURIs` in a `setIsPlaying(false)` I had this same issue when a user would try to play a song that was already playing. Either that or do some kind of conditional where it doesn't try to play something that is already playing and just presents the view.

```Swift
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        
        self.player!.setIsPlaying(false) { (error) -> Void in
            if error != nil {
                print("Error in stopping playback: \(error)")
                return
            }
            
            let options = SPTPlayOptions()
            options.trackIndex = Int32(indexPath.row)
            self.player!.playURIs(self.trackList, withOptions: options, callback: nil)
        }
        
    }
```

**Comment by charleshkang** [comment]
@jakereps Hmm, my issue seems to only happen when a user presses on the back button while any music is playing, because I tested it and it works fine if you pause and then press the back button. 

I'll try wrapping it in a `setIsPlaying` and see what I can do!

**Comment by loretoparisi** [comment]
Hey @charleshkang before pressing the `Back` button when music is playing, did you check if the `SPTAudioStreamingPlaybackDelegate` are not `nil`? Since if you send a `pause` command this will prevent this delegate to fire, but if music is playing, the delegate will continue to fire and send messages to `SPTAudioStreamingController`.

**Comment by charleshkang** [comment]
Hey @loretoparisi, how would I check if `SPTAudioStreamingPlaybackDelegate` is nil? I only set it as a delegate in my `musicPlayerVC` and after that I don't do anything with it. 

That definitely makes sense though!

**Comment by loretoparisi** [comment]
@charleshkang try to remove every `SPTAudioStreamingPlaybackDelegate ` delegates before dismissing the view.

By the way I would do a singleton that responds to these events, this will avoid any issue:

```Objective-c
@interface LPSpotifyHelper()<SPTAudioStreamingDelegate,SPTAudioStreamingPlaybackDelegate> {
    
    SPTSession *session;
    SPTAudioStreamingController *streamingPlayer;
    SPTPlaylistSnapshot *currentPlaylist;
    
    NSMutableArray *tracks;
    NSMutableArray *starredTracks;
    NSMutableArray *savedTracks;
    NSMutableArray *playlists;
    NSArray *partialTracks;
    NSArray *partialPlaylists;
    
    LPTrack *nowPlayingTrack;
    
    dispatch_queue_t taskReadQueue;
    
    LPSpotifyPlayerStatus playerStatus;
    
}

/** Spotify User Session */
@property(nonatomic,strong) SPTSession *session;
/** The current Spotify playlist */
@property(nonatomic,strong) SPTPlaylistSnapshot *currentPlaylist;

@end

@implementation LPSpotifyHelper

@synthesize playlists, tracks, partialPlaylists, partialTracks, starredTracks, savedTracks;
@synthesize currentPlaylist, nowPlayingTrack;
@synthesize startingPlaylistIndex;
@synthesize session;
@synthesize playerStatus;

#pragma mark - Singleton Methods

+(instancetype)sharedInstance {
    static dispatch_once_t pred;
    static LPSpotifyHelper *shared = nil;
    dispatch_once(&pred, ^{
        shared = [[LPSpotifyHelper alloc] init];
    });
    return shared;
}
```

so that you can init once:

```objective-c
/**
 * Initialize the Spotify client
 */
- (void)initSpotify {
    
    [SPTAuth defaultInstance].clientID = kSpotifyClientId;
    [SPTAuth defaultInstance].requestedScopes = [self getAccessTokenScopes];
    [SPTAuth defaultInstance].redirectURL = [NSURL URLWithString:kSpotifyCallbackURL];
    [SPTAuth defaultInstance].tokenRefreshURL = [NSURL URLWithString:kSpotifyTokenRefreshServiceURL];
    [SPTAuth defaultInstance].tokenSwapURL = [NSURL URLWithString:kSpotifySwapURL];
    [SPTAuth defaultInstance].sessionUserDefaultsKey = SpotifyDefaultSessionKey;
    
    if(streamingPlayer == nil) { // initialize the Streaming Player
        
        streamingPlayer = [[SPTAudioStreamingController alloc] initWithClientId:kSpotifyClientId];
        [streamingPlayer setDelegate:self];
        [streamingPlayer setPlaybackDelegate:self];
        [streamingPlayer setTargetBitrate:SPTBitrateHigh callback:^(NSError *error) {
            if(error) {
                [[LPLogger sharedLogger] error:error withMessage:@"Error setting player bitrate"];
            }
        }];
        
        taskReadQueue = dispatch_queue_create(kSpotifyReadTaskQueue, DISPATCH_QUEUE_CONCURRENT);
        dispatch_set_target_queue(taskReadQueue, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0));
        
    }
    
    [self setupCache];
    
}
```

etc.

**Comment by charleshkang** [comment]
Is `LPTrack` and `LPSpotifyPlayerStatus` just two separate classes that you imported?

**Comment by charleshkang** [comment]
@loretoparisi I just tried doing what you did in the above code, but I'm still not sure what I'm doing. I've seen some other sample project using the spotify API and they aren't doing anything like this, but I have a feeling that's because they are around 6-8 month old projects so the framework they are using is older. 

I tried downgrading to an older framework with no luck either.

**Comment by charleshkang** [comment]
@jakereps @loretoparisi 

Just wanted to mention that I got some external help and solved the problem! It was indeed not setting the `SPTAudioStreamingController` to `nil` when dismissing the VC. 

Here's the code snippet that solved everything. 

```- (IBAction)backButtonTapped:(id)sender
{
    if (self.audioPlayer.isPlaying == YES) {
        [self.audioPlayer setIsPlaying:!self.audioPlayer.isPlaying callback:nil];
    }
    self.audioPlayer = nil;
    [self.navigationController popViewControllerAnimated:YES];
}```

I'll keep this issue open just in case!

**Comment by loretoparisi** [comment]
Yes you can do so how setting to nil the player in order to have no delegate called on it, or you can use a singleton. I prefet the latter since having a singleton let you to handle all events (and logic) in one place, share the player through the views, etc.

Sent from Outlook Mobile




On Thu, Mar 3, 2016 at 10:32 AM -0800, "Charles Kang" <notifications@github.com> wrote:












@jakereps @loretoparisi 



Just wanted to mention that I got some external help and solved the problem! It was indeed not setting the SPTAudioStreamingController to nil when dismissing the VC. 



Here's the code snippet that solved everything. 



- (IBAction)backButtonTapped:(id)sender

{

    if (self.audioPlayer.isPlaying == YES) {

        [self.audioPlayer setIsPlaying:!self.audioPlayer.isPlaying callback:nil];

    }

    self.audioPlayer = nil;

    [self.navigationController popViewControllerAnimated:YES];

}



I'll keep this issue open just in case!

**Comment by GabeJacobs** [comment]
Getting this too suddenly.

![Fabric](http://i.imgur.com/mXspcgS.png)

http://i.imgur.com/mXspcgS.png

**Comment by charleshkang** [comment]
@GabeJacobs Suggesting this b/c it's what solved my issue, but are you setting the `SPTAudioStreamingController` to `nil` when dismissing whatever VC it's on?

**Comment by peter9870** [comment]
Same issue here, even after setting `SPTAudioStreamingController` to `nil`

```
override func viewWillDisappear(animated : Bool) {
        super.viewWillDisappear(animated)
        if (self.isMovingFromParentViewController()){
            player = nil
        }
    }
```

**Comment by peter9870** [comment]
Ok, so I fixed this by moving the login flow to AppDelegate, I get a callback when login is successful, which allows me to update my viewcontroller and control music from there. It probably should've already been in AppDelegate but I was being lazy and putting everything in my viewcontroller.

**Comment by cjwhitsitt** [comment]
I'm getting the same stack trace on the crash as @GabeJacobs but I have a singleton that keeps the reference and listens for delegate calls from the SPTAudioStreamingController instead of a VC. I don't believe our issue is the same as the OP's though. @GabeJacobs, can you confirm whether or not your issue was fixed by the suggestion? If not, I'll create a new ticket.

**Comment by GabeJacobs** [comment]
@cjwhitsitt I haven't been able to fix the issue

**Comment by DmytroAnkudinov** [comment]
Similar issue has been fixed in Beta24. @charleshkang  could you please verify that the issue does not reproduce anymore?
